inst/include/lpdens.hpp

```diff
@@ -129,6 +138,36 @@ inline Eigen::VectorXd LPDens1d::kern_gauss(const Eigen::VectorXd& x)
     return x.unaryExpr(f);
 }
 
+//! computes the local bandwidth for a data point; only relevant when a
+//! nearest-neighbor component is specified.
+//! @param x_ev the evaluation point.
+//! @param x the data.
+//! @param bw the bandwidth parameter(s).
+inline double LPDens1d::local_bw(const double& x_ev,
+                                 const Eigen::VectorXd& x,
+                                 const double& bw,
+                                 const double& nn)
+{
+    if (nn > 0.0) {
+        // calculate distances from observations to evaluation point
+        Eigen::VectorXd dists = (x.array() - x_ev).abs();
+
+        // sort in ascending order
+        std::sort(
+            dists.data(),
+            dists.data() + dists.size(),
+            std::less_equal<double>()
+        );
+
+        // calculate index of neighbor such that alpha * n observations are used
+        size_t k = std::lround(nn * static_cast<double>(x.size()));
+
+        return std::max(dists(k), bw);
```

Comment:
Is it quick? The reason that I ask is because selection of the k-th largest can be done in o(n) while sorting is o(nlog(n))

Reply:
https://www.geeksforgeeks.org/quickselect-algorithm/

Reply:
(see quickselect and introselect)